pkg/common/kuadrant_topology.go

```diff
@@ -0,0 +1,385 @@
+			})
+		}
+		if len(index) == 0 {
+			return nil
+		}
+		return index
+	}()
+
+	topologyRepr := struct {
+		Gateways        []Gateway           `json:"gateways"`
+		Routes          []Route             `json:"routes"`
+		Policies        []Policy            `json:"policies"`
+		GatewayPolicies map[string][]string `json:"policiesPerGateway"`
+		PolicyRoute     map[string]string   `json:"policiesTargetingRoutes"`
+		FreeRoutes      map[string][]string `json:"freeRoutesPerGateway"`
+	}{
+		gateways,
+		routes,
+		policies,
+		policiesPerGateway,
+		policiesTargetingRoutes,
+		freeRoutesPerGateway,
+	}
+
+	jsonData, err := json.MarshalIndent(topologyRepr, "", "  ")
+	if err != nil {
+		panic(err)
+	}
+	return string(jsonData)
+}
+
+func buildGatewayPoliciesIndex(t *gatewayAPITopology) map[client.ObjectKey][]KuadrantPolicy {
+	// Build Gateway -> []Policy index with all the policies affecting the indexed gateway
+	index := make(map[client.ObjectKey][]KuadrantPolicy, 0)
+	for _, gatewayNode := range t.GatewaysIndex {
+		// Consisting of:
+		// - Policy targeting directly the gateway
+		// - Policies targeting the descendant routes of the gateway
+		policies := make([]KuadrantPolicy, 0)
+
+		if gatewayNode.DirectPolicy != nil {
+			policies = append(policies, gatewayNode.DirectPolicy.Policy)
+		}
+
+		for _, routeNode := range gatewayNode.Routes {
+			if routeNode.DirectPolicy != nil {
+				policies = append(policies, routeNode.DirectPolicy.Policy)
+			}
+		}
+
+		index[gatewayNode.ObjectKey()] = policies
+	}
+
+	return index
+}
+
+func buildPolicyRouteIndex(t *gatewayAPITopology) map[client.ObjectKey]*gatewayapiv1.HTTPRoute {
+	// Build Policy -> HTTPRoute index with the route targeted by the indexed policy
+	index := make(map[client.ObjectKey]*gatewayapiv1.HTTPRoute, 0)
+	for _, policyNode := range t.PoliciesIndex {
+		if policyNode.TargetedRoute != nil {
+			index[policyNode.ObjectKey()] = policyNode.TargetedRoute.Route
+		}
+	}
+
+	return index
+}
+
+func buildFreeRoutesIndex(t *gatewayAPITopology) map[client.ObjectKey][]*gatewayapiv1.HTTPRoute {
+	// Build Gateway -> []HTTPRoute index with all the routes not targeted by a policy
+	index := make(map[client.ObjectKey][]*gatewayapiv1.HTTPRoute, 0)
+
+	for _, gatewayNode := range t.GatewaysIndex {
+		routes := make([]*gatewayapiv1.HTTPRoute, 0)
+
+		for _, routeNode := range gatewayNode.Routes {
+			if routeNode.DirectPolicy == nil {
+				routes = append(routes, routeNode.Route)
+			}
+		}
+
+		index[gatewayNode.ObjectKey()] = routes
+	}
+
+	return index
+}
+
+type gatewayNode struct {
+	DirectPolicy *policyNode
+	Routes       map[client.ObjectKey]*routeNode
+	Gateway      *gatewayapiv1.Gateway
+}
+
+func (g *gatewayNode) ObjectKey() client.ObjectKey {
+	if g.Gateway == nil {
+		return client.ObjectKey{}
+	}
+
+	return client.ObjectKeyFromObject(g.Gateway)
+}
+
+type routeNode struct {
+	Parents      map[client.ObjectKey]*gatewayNode
+	DirectPolicy *policyNode
```

Comment:
Not that I have a better suggestion for it, but I can see the use of "Direct Policy" becoming confusing, especially in light of the new defs proposed by https://github.com/kubernetes-sigs/gateway-api/pull/2813.

Technically, Kuadrant RLPs and APs are always "Inherited Policies" now, I reckon.

Reply:
After reviewing the implementation of the topolocy and trying to run away from "kuadrant design decisions" to Gateway API topology with policies included, I think we do not need that "direct policy" field in either `routeNode` and `gatewayNode`. 

I will give it a try and come back to this later.